Expose triangle strips in API

AlejandroFernandezLuces opened the pull request:
### Overview
This PR addresses the issue #2340.

I followed the example from `faces` and `lines`, to implement a wrapper for `GetStrip()` and `SetStrip()` functions from `VTK`. 
I also implemented a small assertion to check if the returned array contains any value.

banesullivan commented on pyvista/core/pointset.py:
```diff
@@ -766,6 +766,34 @@ def faces(self, faces):
+        >>> polygon = pv.Rectangle()
+        >>> extruded = polygon.extrude((0, 0, 1))
+        >>> extruded.strips
+        array([4, 0, 1, 4, 5, 4, 1, 2, 5, 6, 4, 2, 3, 6, 7, 4, 3, 0, 7, 4],
+              dtype=int64)
```
The doctest isn't happy with the formatting of this output

banesullivan replied:
```
_______________ [doctest] pyvista.core.pointset.PolyData.strips ________________
774         -------
775         numpy.ndarray
776             Array of strip indices.
777 
778         Examples
779         --------
780         >>> import pyvista as pv
781         >>> polygon = pv.Rectangle()
782         >>> extruded = polygon.extrude((0, 0, 1))
783         >>> extruded.strips
Expected:
    array([4, 0, 1, 4, 5, 4, 1, 2, 5, 6, 4, 2, 3, 6, 7, 4, 3, 0, 7, 4],
          dtype=int64)
Got:
    array([4, 0, 1, 4, 5, 4, 1, 2, 5, 6, 4, 2, 3, 6, 7, 4, 3, 0, 7, 4])
```

adeak replied:
This is probably a windows vs "everything else" thing. On windows int64 is additional info, gets printed. Everywhere else int64 is the default.

AlejandroFernandezLuces replied:
I'm working on a windows machine, I just copied the output from the python output and the test passed in my local machine. I removed that `int64` in commit [a644a86](https://github.com/pyvista/pyvista/pull/3208/commits/a644a86454cf6c8e68ee3c09e13d89faaebd9800).

akaszynski commented:
Good work so far. Tests need to be added to improve coverage. I'd also like to see this added to the `__init__` of `PolyData` with a simple example (ideally with a plot) added.

AlejandroFernandezLuces commented:
> Good work so far. Tests need to be added to improve coverage. I'd also like to see this added to the `__init__` of `PolyData` with a simple example (ideally with a plot) added.

Implemented the `strips` as input in the `__init__`, and also implemented a small test for it, if coverage is still not enough (I'm not sure how to check the diff coverage in local), I'll check tomorrow what lines are not covered, and I'll implement a simple example.

akaszynski commented on examples/02-plot/strips-example.py:
```diff
@@ -0,0 +1,25 @@
+"""
+Build a figure from strips
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Keep this so we can use the sphinx `ref` directive. Also, check title length.

```suggestion
.. _strips_example:

Build a figure from strips
~~~~~~~~~~~~~~~~~~~~~~~~~~
```

AlejandroFernandezLuces replied:
This suggestion `.._strips_example` seems to cause a failure while building the documentation. I'm just following the documentation from other examples, so I'm not sure why this is failing.

adeak replied:
You are missing the space between `..` and `_strips_example`. For future reference, the way @akaszynski posted the suggestion you can click the "commit suggestion" button here to apply the change directly if you want to (then you don't have to copy the change manually).

akaszynski commented on examples/02-plot/strips-example.py:
```diff
@@ -0,0 +1,27 @@
+"""
+.._strips_example:
+
+Strips Figure
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
This is the title shown in the gallery documentation rather than a cell title.
```suggestion
Triangle Strips
~~~~~~~~~~~~~~~
```

akaszynski replied:
Also, you can accept the commit suggestions rather than implementing this in a separate commit (with the commit suggestion button). You can batch accept with:
https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request

AlejandroFernandezLuces commented:
If there is not any more suggestions, I think this is ready from my end, thank you all for your comments! I'll mark the PR as ready for review

adeak commented:
@AlejandroFernandezLuces you've deleted your comment in the meantime that I replied to, so to be on the safe side let me repeat the generally useful parts of it here:

When someone formally posts a "suggestion" in a review comment, you can click the "commit suggestion" button here to apply the change directly if you want to (then you don't have to copy the change manually). See also https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request for further options, e.g. for applying multiple suggestions in a single "batch" via the "files changed" tab of the PR (in this case https://github.com/pyvista/pyvista/pull/3208/files).

akaszynski commented on examples/02-plot/strips-example.py:
```diff
@@ -0,0 +1,27 @@
+Triangle Strips
+~~~~~~~~~~~~~~~
+
+This example shows how to build a simple figure using strips.
```
```suggestion
This example shows how to build a simple :class:`pyvista.PolyData` using triangle strips.

Triangle strips are a more efficient way of storing the connectivity of adjacent triangles.
```

akaszynski commented on examples/02-plot/strips-example.py:
```diff
@@ -0,0 +1,27 @@
+
+# build the mesh
+mesh = pyvista.PolyData(vertices, strips=strips)
+
```
```suggestion
mesh

###############################################################################
# Plot the triangle strips
# ~~~~~~~~~~~~~~~~~~~~~~~~
```

akaszynski commented on pyvista/core/pointset.py:
```diff
@@ -502,6 +502,20 @@ class PolyData(_vtk.vtkPolyData, _PointSet, PolyDataFilters):
         ``PolyData``.
 
+    strips : sequence, optional
+        The strips connectivity array.  Strips require an initial
```
```suggestion
        Triangle strips connectivity array.  Triangle strips require an initial
```

akaszynski commented on pyvista/core/pointset.py:
```diff
@@ -502,6 +502,20 @@ class PolyData(_vtk.vtkPolyData, _PointSet, PolyDataFilters):
+        The strips connectivity array.  Strips require an initial
+        triangle, and the following points of the strip. Each
+        triangle is built with the new point and the two previous
+        points. As ``lines`` and ``faces``, this array requires a
```
```suggestion
        points. Just as in ``lines`` and ``faces``, this array requires a
```

akaszynski commented on pyvista/core/pointset.py:
```diff
@@ -502,6 +502,20 @@ class PolyData(_vtk.vtkPolyData, _PointSet, PolyDataFilters):
+        triangle, and the following points of the strip. Each
+        triangle is built with the new point and the two previous
+        points. As ``lines`` and ``faces``, this array requires a
+        padding indicating the number of points. For example,
+        triangle ``[0, 1, 3]`` and the next point ``[2]`` will be
+        represented as ``[4, 0, 1, 3, 2]``.
```
```suggestion
        padding indicating the number of points. For example,
        a single triangle strip of ``[0, 1, 2, 3, 6, 7, 4, 5, 0, 1]`` requires padding of
        ``10`` and should input as ``[10, 0, 1, 2, 3, 6, 7, 4, 5, 0, 1]``.
```

akaszynski commented on pyvista/core/pointset.py:
```diff
@@ -550,6 +564,11 @@ class PolyData(_vtk.vtkPolyData, _PointSet, PolyDataFilters):
     >>> lines = np.hstack([[2, 0, 1], [2, 1, 2]])
     >>> mesh = pyvista.PolyData(vertices, lines=lines)
 
+    Initialize from vertices and strips
```
```suggestion
    Initialize from vertices and triangle strips.
```

akaszynski commented on tests/test_polydata.py:
```diff
@@ -792,6 +792,20 @@ def test_lines():
+    vertices = np.array([[0, 0, 0], [1, 0, 0], [1, 0.5, 0], [0, 0.5, 0]])
+    strips = np.hstack([[4, 0, 1, 3, 2]])
+    strips_init = pyvista.PolyData(vertices, strips=strips)
+    assert len(strips_init.strips) == 5
```
```suggestion
    assert len(strips_init.strips) == len(strips)
```

akaszynski commented on tests/test_polydata.py:
```diff
@@ -792,6 +792,20 @@ def test_lines():
+
+    # init with strips test
+    vertices = np.array([[0, 0, 0], [1, 0, 0], [1, 0.5, 0], [0, 0.5, 0]])
+    strips = np.hstack([[4, 0, 1, 3, 2]])
```
No reason to have `hstack` here.
```suggestion
    strips = np.array([4, 0, 1, 3, 2])
```

akaszynski commented on tests/test_polydata.py:
```diff
@@ -792,6 +792,20 @@ def test_lines():
+    # add strips using the setter
+    strips_setter = pyvista.PolyData(vertices)
+    strips_setter.strips = strips
+    assert len(strips_setter.strips) == 5
```
```suggestion
    assert len(strips_setter.strips) == len(strips)
```

akaszynski commented:
Making good progress and test coverage is good. We're missing two things:
- Add `n_strips` from [GetNumberOfStrips](https://vtk.org/doc/nightly/html/classvtkPolyData.html#a5efd013787c3d8993e7753a04a2133a2)
- Update the `repr` to include the number of strips.

AlejandroFernandezLuces commented:
> Making good progress and test coverage is good. Close, just missing two things:
> 
> * Add `n_strips` from [GetNumberOfStrips](https://vtk.org/doc/nightly/html/classvtkPolyData.html#a5efd013787c3d8993e7753a04a2133a2)
> * Update the `repr` to include the number of strips.

Both done, let me know if there's anything else that needs to be done :)

akaszynski commented on pyvista/core/dataset.py:
```diff
@@ -2019,6 +2019,9 @@ def _get_attrs(self):
         attrs = []
         attrs.append(("N Cells", self.GetNumberOfCells(), "{}"))
         attrs.append(("N Points", self.GetNumberOfPoints(), "{}"))
+        has_strips = getattr(self, "GetNumberOfStrips", None)
+        if callable(has_strips):
+            attrs.append(("N Strips", self.GetNumberOfStrips(), "{}"))
```
I'd probably just check if it's an instance here with:
```suggestion
        if isinstance(self, pyvista.PolyData):
            attrs.append(("N Strips", self.n_strips, "{}"))
```

This is because it's guaranteed to have this API.

akaszynski reviewed (approved):
I made substantial changes to the example to make it clearer how triangle strips were being created. Otherwise, LGTM.

![tmp](https://user-images.githubusercontent.com/11981631/187298796-c999d24e-4250-4599-bab9-2383a68e6027.png)

banesullivan reviewed (approved):
This turned out really excellent!
